### PR TITLE
BFB-557: Tutorial System

### PR DESCRIPTION
**Pull request by m4rtint**
**JIRA Ticket**
https://perigongames.atlassian.net/browse/BFB-557

**Description**
- Added Carousel
- Added Tutorial System

NOTE - None of the tutorials has any content right now.

**Screenshots / GIFs**
https://github.com/PerigonGames/BForBoss/assets/5779541/9d3869a0-51d0-45e2-b4a6-a8863efd9350

### CODE REVIEW

File: Assets/_BForBoss/_Core/Scripts/UserInterface/Tutorial/TutorialViewsManager.cs [hunk 1]
```diff
@@ -0,0 +1,88 @@
...
+        }
+
+        private void Start()
+        {
+            SetupTutorials();
+            _energyTutorialView.Initialize();
+            _basicTutorialView.Initialize();
+        }
+
+        private IStateManager StateManager => BForBoss.StateManager.Instance;
```
Comment (Gursaanj) [inline Assets/_BForBoss/_Core/Scripts/UserInterface/Tutorial/TutorialViewsManager.cs:36]: Getters before methods

File: Assets/_BForBoss/_Core/Scripts/UserInterface/Tutorial/TutorialViewsManager.cs [hunk 2]
```diff
@@ -0,0 +1,88 @@
...
+
+        public void Show(TutorialState state)
```
Comment (Gursaanj) [inline Assets/_BForBoss/_Core/Scripts/UserInterface/Tutorial/TutorialViewsManager.cs:38]: Public methods before private ones

File: Assets/_BForBoss/_UICore/Scripts/CarouselPanelView.cs [hunk 3]
```diff
@@ -0,0 +1,65 @@
...
+
+            if (!isBackShown)
+            {
+                _backButton.gameObject.SetActive(false);
+            }
+        }
+
+        private void Awake()
+        {
+            if (_backButton == null)
```
Comment (Gursaanj) [inline Assets/_BForBoss/_UICore/Scripts/CarouselPanelView.cs:34]: can use the new PanicIfNull method.

File: Assets/_BForBoss/_UICore/Scripts/CarouselPanelView.cs [hunk 4]
```diff
@@ -0,0 +1,65 @@
...
+            {
+                PanicHelper.Panic(new Exception("Continue Button missing from carousel panel view"));
+            }
+
+            if (_exitButton == null)
+            {
+                PanicHelper.Panic(new Exception("Exit Button missing from carousel panel view"));
+            }
+            
+            _backButton.onClick.AddListener(() =>
```
Comment (Gursaanj) [inline Assets/_BForBoss/_UICore/Scripts/CarouselPanelView.cs:49]: probably overkill, but `OnDestroy` you can removeListeners

File: Assets/_BForBoss/_UICore/Scripts/CarouselView.cs [hunk 5]
```diff
@@ -0,0 +1,104 @@
...
+                _carouselPanels[index].gameObject.SetActive(false);
+            });
+            SetupPanels();
+            SetupButtons();
+        }
+
+        private void Awake()
+        {
+            _carouselPanels = gameObject.GetComponentsInChildren<CarouselPanelView>(includeInactive: true);
+            if (_carouselPanels.IsNullOrEmpty())
```
Comment (Gursaanj) [inline Assets/_BForBoss/_UICore/Scripts/CarouselView.cs:37]: Can use the new `PanicIfNullOrEmpty` method

**Review by Gursaanj — COMMENTED** [review]
Looks good, small nitpick comment only

Just want to make sure Pausing the game while the Tutorial are up, does not break anything.

**Comment by m4rtint** [comment]
> Looks good, small nitpick comment only
> 
> Just want to make sure Pausing the game while the Tutorial are up, does not break anything.

You shouldn't be able to pause when tutorial is up. 

The only way the game is paused:

https://github.com/PerigonGames/BForBoss/blob/34a679a166c6c9a63efc7672c91167af1a676313/Assets/_BForBoss/_Core/Scripts/Managers/BaseWorldManager.cs#L157

File: Assets/_BForBoss/_UICore/Scripts/CarouselPanelView.cs (hunk 4 above)
Comment (m4rtint) [reply Assets/_BForBoss/_UICore/Scripts/CarouselPanelView.cs:38]: Removed listeners

File: Assets/_BForBoss/_Core/Scripts/UserInterface/Tutorial/TutorialViewsManager.cs (hunk 2 above)
Comment (m4rtint) [reply Assets/_BForBoss/_Core/Scripts/UserInterface/Tutorial/TutorialViewsManager.cs:27]: Reorganized

**Review by Gursaanj — APPROVED** [review]

**Pull request merged**